### PR TITLE
ci: optimize github action

### PR DESCRIPTION
**Pull request by ben1009**

### CODE REVIEW

**Comment by ben1009** [comment]
seems alright
[first run]([ci: optimize github action & fix clippy warnings](https://github.com/ben1009/cargo-make/actions/runs/7445347925))
image
[with cache hit
]([rm commented code](https://github.com/ben1009/cargo-make/actions/runs/7445680553))image

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -1,7 +1,13 @@
 name: CI
 on: [push, pull_request]
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
```
Comment (sagiegurari) [inline .github/workflows/ci.yml:5]: can you just explain me what this does? already all jobs run in parallel so what does this solve?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -1,7 +1,13 @@
...
+
 env:
   CLICOLOR_FORCE: 1
+  CARGO_REGISTRIES_CRATES_IO_PROTOCOL: sparse
```
Comment (sagiegurari) [inline .github/workflows/ci.yml:9]: nice

**Comment by sagiegurari** [comment]
@ben1009 thanks for this, I found a critical cargo change that made me fix something and already stole some of the stuff you did in this PR (but not all), so it requires conflict resolving and if possible can you answer questions i wrote?

File: src/lib/mod.rs [hunk 3]
```diff
@@ -49,7 +49,6 @@
     indirect_structural_match,
     ineffective_unstable_trait_impl,
     inline_no_sanitize,
-    invalid_alignment,
```
Comment (sagiegurari) [inline src/lib/mod.rs:52]: updated the rust lint rules, i'm nnot sure this change is needed anymore

File: src/lib/mod.rs [hunk 4]
```diff
@@ -90,7 +89,6 @@
     path_statements,
     patterns_in_fns_without_body,
     pointer_structural_match,
-    private_in_public,
```
Comment (sagiegurari) [inline src/lib/mod.rs:93]: updated the rust lint rules, i'm nnot sure this change is needed anymore

File: .github/workflows/ci.yml (hunk 1 above)
Comment (ben1009) [reply .github/workflows/ci.yml:5]: it means if same pr committed multiple times, will just cancel the ongoing test, if have any,just run the latest one instead

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -17,12 +23,13 @@ jobs:
...
-    - name: Install cargo-make
-      run: cargo install --debug cargo-make
+        toolchain: ${{ matrix.rust }}    
+    - run: cargo install --debug cargo-make
+    - name: Rust Cache
+      uses: Swatinem/rust-cache@v2
     - name: Run CI
       if: matrix.rust == 'stable' || matrix.rust == 'beta' || matrix.rust == 'nightly'
-      run: cargo run --bin cargo-make --target-dir target/ci -- make ci-flow
+      run: cargo run --bin cargo-make --target-dir target/ci -- make ci-flow
```
Comment (sagiegurari) [inline .github/workflows/ci.yml:32]: change can be removed

File: src/lib/mod.rs [hunk 6]
```diff
@@ -54,7 +54,6 @@
     indirect_structural_match,
     ineffective_unstable_trait_impl,
     inline_no_sanitize,
-    internal_features,
```
Comment (sagiegurari) [inline src/lib/mod.rs:57]: i ran with nightly and i don't see issues. why remove?

File: src/lib/mod.rs [hunk 7]
```diff
@@ -100,8 +99,6 @@
     path_statements,
     patterns_in_fns_without_body,
     pointer_structural_match,
-    private_bounds,
-    private_interfaces,
```
Comment (sagiegurari) [inline src/lib/mod.rs:104]: i ran with nightly and i don't see issues. why remove?

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -17,12 +23,13 @@ jobs:
     - name: Install rust
       uses: dtolnay/rust-toolchain@master
       with:
-        toolchain: ${{ matrix.rust }}
-    - name: Install cargo-make
-      run: cargo install --debug cargo-make
+        toolchain: ${{ matrix.rust }}    
+    - run: cargo install --debug cargo-make
```
Comment (sagiegurari) [inline .github/workflows/ci.yml:27]: i think we can drop this change, right?

**Review by sagiegurari — CHANGES_REQUESTED** [review]

**Comment by sagiegurari** [comment]
merged. thanks a lot @ben1009

**Pull request merged**